user settings: Add setting to make login notifications optional.

pragatiagrawal31 opened the pull request:
This adds a feature under "Notification" section of "Settings" tab,
which lets user enable or disable login emails notification.

Fixes:#5854

**GIFs or Screenshots:** 
![image](https://user-images.githubusercontent.com/40331304/44568429-636ebc00-a794-11e8-9733-f818be3b959f.png)

timabbott commented:
@rishig thoughts on what we should call this option?  Maybe "Email notifications on new logins."?

timabbott commented:
@pragatiagrawal31 generally this looks great; the big thing that's missing is that there should be a test, probably in `SendLoginEmailTest`, that verifies that with this setting disabled, the emails don't get sent.

pragatiagrawal31 commented:
That makes sense, I’ll be adding new tests soon. Yeah, “Email notifications on new logins.” sounds better and if this is final, I’ll change this accordingly?

rishig commented:
How about "Send email notifications for new logins to my account"?

timabbott commented:
That sounds good to me!

pragatiagrawal31 commented:
I've removed the case when the condition is `enable_email_login` is true as it's causing the tests failed. The true condition is already been tested in `test_send_login_emails_if_send_login_email_setting_is_true` as the default value is true. But I'm still not able to figure out why previous tests are failing after much debugging. 
I've removed below-mentioned lines which were causing the test's fail.
```
do_change_notification_settings(user, "enable_login_emails", True)
        self.login(user.email)
        self.assertTrue(user.enable_login_emails)
        self.assertEqual(len(mail.outbox), 1)
```
@shubhamdhama and I have tried setting up the circleci debugging but can't do that successfully

shubhamdhama commented:
Today I did more debugging for this and tried various ways here but I think this error is in master as well as this test also don't pass there:
```

    @override_settings(SEND_LOGIN_EMAILS=True)
    def test_enable_login_emails_user_setting_is_true(self) -> None:
        self.assertTrue(settings.SEND_LOGIN_EMAILS)
        user = self.example_user('hamlet')
        password = initial_password(user.email)
        self.client_post("/accounts/login/", info={"username": user.email, "password": password})
        self.assertEqual(len(mail.outbox), 1)
```

shubhamdhama commented:
Ahhhh... I think I've got what's happening here.
https://github.com/zulip/zulip/blob/37acb3e2cbe58a83795fdf7d3251a4cbe1f811fb/zerver/signals.py#L70-L72
When we run tests locally, we have already created the user from some time so this condition don't get true and return isn't executed, hence we get the positive tests but in test suite it's threshold isn't surpassed and tests fails. So I think @pragatiagrawal31 you need to do some mocking here, you can take some help from `test_send_login_emails_if_send_login_email_setting_is_true`

timabbott commented:
Ahh, yeah, that makes sense.

pragatiagrawal31 commented:
Thankyou @shubhamdhama. I have done mocking and updated the tests accordingly

timabbott commented:
Merged, after renumbering the migration and changing the issue this closes to #5795, since there's more we'll want to do to resolve #5854.  See ba4e4e38aeaa708d9ed7be3f218e5ab6d61ac7e6 for the final version.

Thanks @pragatiagrawal31!  I think a good next step would be to add an "unsubscribe" link to the "new login" notification emails, that also toggles this flag.  We have similar code around functions like `one_click_unsubscribe_link` in `zerver/lib/notifications.py`.  Make sure you read https://zulip.readthedocs.io/en/latest/subsystems/email.html before starting on that :)

pragatiagrawal31 commented:
I’m looking forward to work on adding the “unsubscribe” link to the “new login” emails. Thankyou @timabbott ! Surely, will go through the docs :)
